### PR TITLE
Send ccd

### PR DESCRIPTION
**Pull request by shjortConcordium**
## Purpose

Allow sending Ccd's from inside the wallet

## Changes

- Added the send ccd flow.
- Changed the UI in the sendTransaction (from dApp).
- Some small refactorings.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Comment by orhoj** [comment]
If I enter something invalid in both input fields, then I end up with a scrollbar. I think the UI should be so that that cannot happen.

**Comment by orhoj** [comment]
The input fields width should be similar to how it is in Figma.

**Comment by orhoj** [comment]
The "Cancel" and "Confirm" buttons should be styled closer to how it is in Figma.

**Comment by orhoj** [comment]
If I send a transaction and am on the "Finish" view, then if I click the action button for sending, then the input fields are still pre-filled with my previous transaction. That seems a little odd.

**Comment by orhoj** [comment]
In Figma the input field has the CCD symbol in front of the amount you input. If it's easy to add, then we should also have it here.

File: packages/browser-wallet/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,11 @@
 
 ## Unreleased
 
+### Added
+
+-   Toasts for showing errors when failing to retrieve account balance or transaction history.
+-   Send ccd page.
```
Comment (orhoj) [inline packages/browser-wallet/CHANGELOG.md:8]:
```suggestion
-   Send CCD page.
```

File: packages/browser-wallet/src/popup/pages/Account/Account.tsx [hunk 2]
```diff
@@ -73,7 +76,11 @@ function Account({
                             <ConnectedBox setDetailsExpanded={setDetailsExpanded} />
                         </div>
                         <div className="account-page__routes">
-                            {isConfirmed && <Outlet />}
+                            {isConfirmed && (
+                                <AccountContext account={selectedCred}>
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/Account.tsx:80]: I am not sure if it works with the context here.

File: packages/browser-wallet/src/popup/shared/i18n/da.ts [hunk 3]
```diff
@@ -15,6 +15,22 @@ const t: typeof en = {
...
+        },
+        ccdAmount: {
+            required: 'Skriv et beløb',
+            invalid: 'Invalid beløb',
+            insufficient: 'Ikke nok penge',
+            zero: 'Beløbet må ikke være nul',
+        },
+    },
+    transactionReceipt: {
+        sender: 'Sendende konti',
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/shared/i18n/da.ts:31]:
```suggestion
        sender: 'Afsender',
```

File: packages/browser-wallet/src/popup/shared/i18n/da.ts [hunk 4]
```diff
@@ -15,6 +15,22 @@ const t: typeof en = {
...
+        cost: 'Estimeret transaktions afgift',
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/shared/i18n/da.ts:32]:
```suggestion
        cost: 'Estimerede transaktionsomkostninger',
```

File: packages/browser-wallet/src/popup/shared/i18n/da.ts [hunk 5]
```diff
@@ -15,6 +15,22 @@ const t: typeof en = {
         confirmed: 'Godkendt af',
         rejected: 'Afvist af',
     },
+    utils: {
+        address: {
+            required: 'Skriv en addresse',
+            invalid: 'Invalid addresse',
+        },
+        ccdAmount: {
+            required: 'Skriv et beløb',
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/shared/i18n/da.ts:24]:
```suggestion
            required: 'Indtast et beløb',
```

File: packages/browser-wallet/src/popup/shared/i18n/da.ts [hunk 6]
```diff
@@ -15,6 +15,22 @@ const t: typeof en = {
         confirmed: 'Godkendt af',
         rejected: 'Afvist af',
     },
+    utils: {
+        address: {
+            required: 'Skriv en addresse',
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/shared/i18n/da.ts:20]:
```suggestion
            required: 'Indtast en addresse',
```

File: packages/browser-wallet/src/popup/shared/i18n/da.ts [hunk 7]
```diff
@@ -15,6 +15,22 @@ const t: typeof en = {
...
+            invalid: 'Invalid addresse',
+        },
+        ccdAmount: {
+            required: 'Skriv et beløb',
+            invalid: 'Invalid beløb',
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/shared/i18n/da.ts:25]:
```suggestion
            invalid: 'Ugyldigt beløb',
```

File: packages/browser-wallet/src/popup/shared/i18n/da.ts [hunk 8]
```diff
@@ -15,6 +15,22 @@ const t: typeof en = {
         confirmed: 'Godkendt af',
         rejected: 'Afvist af',
     },
+    utils: {
+        address: {
+            required: 'Skriv en addresse',
+            invalid: 'Invalid addresse',
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/shared/i18n/da.ts:21]:
```suggestion
            invalid: 'Ugyldig addresse',
```

File: packages/browser-wallet/src/popup/shared/TransactionReceipt/TransactionReceipt.scss [hunk 9]
```diff
@@ -0,0 +1,59 @@
+$transaction-receipt-horizontal-padding: rem(20px);
+
+.transaction-receipt {
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    background-color: $color-off-white;
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/shared/TransactionReceipt/TransactionReceipt.scss:7]: What about dark mode?

File: packages/browser-wallet/src/popup/shared/TransactionReceipt/DisplayCost.tsx [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+import clsx from 'clsx';
+import { displayAsCcd } from 'wallet-common-helpers';
+import { useTranslation } from 'react-i18next';
+
+type Props = {
+    cost: bigint;
+    className?: string;
+};
+
+export default function DisplayAddress({ cost, className }: Props) {
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/shared/TransactionReceipt/DisplayCost.tsx:11]:
```suggestion
export default function DisplayCost({ cost, className }: Props) {
```

File: packages/browser-wallet/src/popup/pages/Account/i18n/da.ts [hunk 11]
```diff
@@ -34,6 +34,21 @@ const t: typeof en = {
...
+        },
+        buttons: {
+            cancel: 'Annuller',
+            confirm: 'Bekræft',
+            finish: 'Færdigør',
+            continue: 'Fortsæt',
+        },
+        title: 'Send CCD',
+        receiptTitle: 'Transaktion: Send CCD',
+        fee: 'Estimated transaction fee',
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/i18n/da.ts:50]: English

File: packages/browser-wallet/src/popup/pages/Account/i18n/da.ts [hunk 12]
```diff
@@ -34,6 +34,21 @@ const t: typeof en = {
             done: 'Færdig',
         },
     },
+    sendCcd: {
+        labels: {
+            ccd: 'Skriv et beløb at overføre',
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/i18n/da.ts:39]:
```suggestion
            ccd: 'Indtast et beløb at overføre',
```

File: packages/browser-wallet/src/popup/pages/SendTransaction/SendTransaction.tsx [hunk 13]
```diff
@@ -88,33 +78,30 @@ export default function SendTransaction({ onSubmit, onReject }: Props) {
...
-            <h5>{t('sender')}:</h5>
-            <p className="send-transaction__address">{accountAddress}</p>
-            {transactionType === AccountTransactionType.SimpleTransfer && <DisplaySimpleTransfer payload={payload} />}
-            {transactionType === AccountTransactionType.UpdateSmartContractInstance && (
-                <DisplayUpdateContract payload={payload} parameters={state.payload.parameters} />
-            )}
-            {transactionType === AccountTransactionType.InitializeSmartContractInstance && (
-                <DisplayInitContract payload={payload} parameters={state.payload.parameters} />
-            )}
+            <TransactionReceipt title="Sign transaction" sender={accountAddress}>
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/SendTransaction/SendTransaction.tsx:87]: Translation of the title is missing.

File: packages/browser-wallet/src/popup/pages/Account/SendCcd/SendCcd.tsx [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+import ConfirmTransfer from './ConfirmTransfer';
+import { routes } from './routes';
+
+interface Props {
+    setDetailsExpanded: (expanded: boolean) => void;
+}
+
+export default function SendCcd({ setDetailsExpanded }: Props) {
+    const [payload, setPayload] = useState<SimpleTransferPayload>();
+    const cost = useAsyncMemo(getSimpleTransferCost, noOp, []);
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/SendCcd/SendCcd.tsx:16]: I don't get an error message if the cost could not be fetched, it is just shown as `0`. I guess it doesn't make sense to allow sending a transaction if we don't have the cost available when building the transaction?

File: packages/browser-wallet/src/popup/pages/Account/SendCcd/CreateTransfer.tsx [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    const nav = useNavigate();
+    const { accountInfo } = useContext<AccountContextValues>(accountContext);
+
+    const validateAmount: Validate<string> = (amount) => validateTransferAmount(amount, accountInfo, cost);
+
+    if (!address) {
+        return null;
+    }
+
+    const onSubmit: SubmitHandler<FormValues> = async (vs) => {
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/SendCcd/CreateTransfer.tsx:45]: I don't think this should be `async`.

File: packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+import { usePrivateKey } from '@popup/shared/utils/account-helpers';
+import Button from '@popup/shared/Button';
+import DisplaySimpleTransfer from '@popup/shared/TransactionReceipt/displayPayload/DisplaySimpleTransfer';
+import { useNavigate } from 'react-router-dom';
+import { absoluteRoutes } from '@popup/constants/routes';
+import TransactionReceipt from '@popup/shared/TransactionReceipt/TransactionReceipt';
+import { routes } from './routes';
+
+interface Props {
+    payload: SimpleTransferPayload | undefined;
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx:17]: Why would the payload ever be `undefined` when this page is shown? In that case there would be nothing to confirm.

File: packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        setDetailsExpanded(false);
+        return () => setDetailsExpanded(true);
+    }, []);
+
+    const send = async () => {
+        if (!payload) {
+            throw new Error('Missing payload');
+        }
+        if (!address || !key) {
+            throw new Error('Missing address for selected account');
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx:40]: Or `key`.

File: packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        const nonce = await client.getNextAccountNonce(new AccountAddress(address));
+        if (!nonce) {
+            throw new Error('No nonce found for sender');
+        }
+        const header = {
+            // TODO: add better default?
+            expiry: new TransactionExpiry(new Date(Date.now() + 3600000)),
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx:48]: I believe we have this TODO and this default in more than one place now. Make a function for getting the default transaction expiry, and then we can update it globally if we decide on another value.

File: packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    }, []);
+
+    const send = async () => {
+        if (!payload) {
+            throw new Error('Missing payload');
+        }
+        if (!address || !key) {
+            throw new Error('Missing address for selected account');
+        }
+        const nonce = await client.getNextAccountNonce(new AccountAddress(address));
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx:42]: You do `new AccountAddress(address)` multiple times. Make a variable.

File: packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+            throw new Error('No nonce found for sender');
+        }
+        const header = {
+            // TODO: add better default?
+            expiry: new TransactionExpiry(new Date(Date.now() + 3600000)),
+            sender: new AccountAddress(address),
+            nonce: nonce.nonce,
+        };
+        const transaction = { payload, header, type: AccountTransactionType.SimpleTransfer };
+        sendTransaction(client, transaction, key).then(setHash);
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx:53]: You are using `await` syntax elsewhere in this `async` function, so I think you should stick to that here as well. So:
```
const transactionHash = await sendTransaction(...);
setHash(transactionHash);
```

**Review by orhoj — CHANGES_REQUESTED** [review]

File: packages/browser-wallet/src/popup/pages/Account/SendCcd/SendCcd.tsx (hunk 14 above)
Comment (shjortConcordium) [reply packages/browser-wallet/src/popup/pages/Account/SendCcd/SendCcd.tsx:16]: :thinking:
Yeah I guess we might want to disable continuing or something.

File: packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx (hunk 16 above)
Comment (shjortConcordium) [reply packages/browser-wallet/src/popup/pages/Account/SendCcd/ConfirmTransfer.tsx:17]: It wouldn't, it was just "lazy" typing. I have refactored this away though.

**Comment by orhoj** [comment]
We need the UI changes we discussed with Jens.

**Comment by orhoj** [comment]
The buttons are still missing margin if I collapse the balance view.

**Review by orhoj — APPROVED** [review]

**Pull request merged**